idaes/core/base/unit_model.py

```diff
@@ -526,14 +526,19 @@ def initialize(blk, *args, **kwargs):
             c.deactivate()
 
         # Remember to collect flags for fixed vars
-        flags = blk.initialize_build(*args, **kwargs)
-
-        # If costing block exists, activate and initialize
-        for c in init_order:
-            c.activate()
-
-            if hasattr(c, "initialize"):
-                c.initialize(**cost_args)
+        try:
+            flags = blk.initialize_build(*args, **kwargs)
+        except InitializationError:
+            raise
+        finally:
+            # In case the InitializationError is caught,
+            # leave the unit unmodified
+            # If costing block exists, activate and initialize
+            for c in init_order:
```

Comment:
We definitely want to leave the model in an unmodified state if an InitializationError is encountered, but I think this issue is somewhat broader than this.

1. We probably need to update control volumes and state blocks to do the same checks to make sure they reset themselves after an initialization failure. Control volumes are easy enough, but state blocks would have to be done on an individual basis (for now).
2. Similarly, costing blocks would also need to handle initialization failures gracefully.
3. Finally, does it make sense to try to initialize the costing block given that the unit model failed? I can see arguments both ways.

Note that we intend to address many of these in a new initialization API that is currently in the planning stages.